### ISSUE TITLE
建议下一个版本支持一下mac，花了我好长时间来适配mac

### ISSUE THREAD

**Issue by yyaadet** [MoonshotAI/Kimi-Audio#15]
mac没有cuda版本的flash attn，下个版本移除吧。代码我修改好了。

➜  Kimi-Audio git:(master) ✗ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   kimia_infer/models/detokenizer/flow_matching/dit_block.py
        modified:   kimia_infer/models/tokenizer/glm4 (untracked content)
        modified:   kimia_infer/models/tokenizer/whisper_Lv3/modeling_whisper.py
        modified:   requirements.txt

diff files:

diff --git a/kimia_infer/models/detokenizer/flow_matching/dit_block.py b/kimia_infer/models/detokenizer/flow_matching/dit_block.py
index f60c1d3..b8b20fc 100644
--- a/kimia_infer/models/detokenizer/flow_matching/dit_block.py
+++ b/kimia_infer/models/detokenizer/flow_matching/dit_block.py
@@ -1,13 +1,7 @@
-import torch
-import torch.nn as nn
-
-
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
 
-from flash_attn import flash_attn_varlen_func, flash_attn_varlen_qkvpacked_func
-
 
 def reshape_for_broadcast(freqs_cis: torch.Tensor, x: torch.Tensor):
     # x shape: bsz, seqlen, self.n_local_heads, self.head_hidden_dim / 2
@@ -50,14 +44,14 @@ class Attention(nn.Module):
         attn_drop: float = 0.0,
         proj_drop: float = 0.0,
         norm_layer: nn.Module = nn.LayerNorm,
-        flash_attention: bool = True,
+        flash_attention: bool = False,
     ) -> None:
         super().__init__()
         assert dim % num_heads == 0, "dim should be divisible by num_heads"
         self.num_heads = num_heads
         self.head_dim = dim // num_heads
         self.scale = self.head_dim**-0.5
-        self.fused_attn = flash_attention
+        self.fused_attn = False
 
         self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
         self.qk_norm = qk_norm
@@ -80,106 +74,46 @@ class Attention(nn.Module):
         nopadding=True,
     ) -> torch.Tensor:
         B, N, C = x.shape
-
-        if self.fused_attn:
-            if nopadding:
-                qkv = self.qkv(x)
-                qkv = qkv.view(B * N, self.num_heads * 3, self.head_dim)
-                q, k, v = qkv.split([self.num_heads] * 3, dim=1)
-                q, k = self.q_norm(q), self.k_norm(k)
-
-                q = q.view(B, N, self.num_heads, self.head_dim)
-                k = k.view(B, N, self.num_heads, self.head_dim)
-                v = v.view(B, N, self.num_heads, self.head_dim)
-
-                if rotary_pos_emb is not None:
-                    q, k = apply_rotary_emb(q, k, rotary_pos_emb)
-
-                if incremental_state is not None:
-                    if "prev_k" in incremental_state:
-                        prev_k = incremental_state["prev_k"]
-                        k = torch.cat([prev_k, k], dim=1)
-
-                    if "cur_k" not in incremental_state:
-                        incremental_state["cur_k"] = {}
-                    incremental_state["cur_k"] = k
-
-                    if "prev_v" in incremental_state:
-                        prev_v = incremental_state["prev_v"]
-                        v = torch.cat([prev_v, v], dim=1)
-
-                    if "cur_v" not in incremental_state:
-                        incremental_state["cur_v"] = {}
-                    incremental_state["cur_v"] = v
-
-                q = q.view(B * N, self.num_heads, self.head_dim)
-                k = k.view(-1, self.num_heads, self.head_dim)
-                v = v.view(-1, self.num_heads, self.head_dim)
-
-                x = flash_attn_varlen_func(
-                    q=q,
-                    k=k,
-                    v=v,
-                    cu_seqlens_q=cu_seqlens,
-                    cu_seqlens_k=cu_seqlens_k,
-                    max_seqlen_q=max_seqlen,
-                    max_seqlen_k=max_seqlen_k,
-                    dropout_p=self.attn_drop.p if self.training else 0.0,
-                )
-            else:
-
-                if incremental_state is not None:
-                    raise NotImplementedError(
-                        "It is designed for batching inference. AR-chunk is not supported currently."
-                    )
-
-                qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, self.head_dim)
-                if self.qk_norm:
-                    q, k, v = qkv.unbind(2)
-                    q, k = self.q_norm(q), self.k_norm(k)
-                    # re-bind
-                    qkv = torch.stack((q, k, v), dim=2)
-
-                # pack qkv with seq_len
-                qkv_collect = []
-                for i in range(qkv.shape[0]):
-                    qkv_collect.append(qkv[i, : seq_len[i], :, :, :])
-
-                qkv = torch.cat(qkv_collect, dim=0)
-
-                x = flash_attn_varlen_qkvpacked_func(
-                    qkv=qkv,
-                    cu_seqlens=cu_seqlens,
-                    max_seqlen=max_seqlen,
-                    dropout_p=self.attn_drop.p if self.training else 0.0,
-                )
-
-                # unpack and pad 0
-                x_collect = []
-                for i in range(B):
-                    x_collect.append(x[cu_seqlens[i] : cu_seqlens[i + 1], :, :])
-                x = torch.nn.utils.rnn.pad_sequence(
-                    x_collect, batch_first=True, padding_value=0
-                )
-
-        else:
-            q = q * self.scale
-            attn = q @ k.transpose(-2, -1)
-            attn = attn.softmax(dim=-1)
-            attn = self.attn_drop(attn)
-            x = attn @ v
-            x = x.transpose(1, 2)
-
-        x = x.reshape(B, N, C)
+        
+        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, self.head_dim).permute(2, 0, 1, 3, 4)
+        q, k, v = qkv[0], qkv[1], qkv[2]  # make torchscript happy (cannot use tensor as tuple)
+        
+        if self.qk_norm:
+            q = self.q_norm(q)
+            k = self.k_norm(k)
+            
+        if rotary_pos_emb is not None:
+            q, k = apply_rotary_emb(q, k, rotary_pos_emb)
+            
+        if incremental_state is not None:
+            if "prev_k" in incremental_state:
+                prev_k = incremental_state["prev_k"]
+                k = torch.cat([prev_k, k], dim=1)
+
+            if "cur_k" not in incremental_state:
+                incremental_state["cur_k"] = {}
+            incremental_state["cur_k"] = k
+
+            if "prev_v" in incremental_state:
+                prev_v = incremental_state["prev_v"]
+                v = torch.cat([prev_v, v], dim=1)
+
+            if "cur_v" not in incremental_state:
+                incremental_state["cur_v"] = {}
+            incremental_state["cur_v"] = v
+            
+        q = q * self.scale
+        attn = (q @ k.transpose(-2, -1))
+        
+        attn = attn.softmax(dim=-1)
+        attn = self.attn_drop(attn)
+        
+        x = (attn @ v).transpose(1, 2).reshape(B, N, C)
         x = self.proj(x)
         x = self.proj_drop(x)
         return x
 
 
-def modulate(x, shift, scale):
-    return x * (1 + scale) + shift
-
-
 class FinalLayer(nn.Module):
     """
     The final layer of DiT.
@@ -295,3 +229,7 @@ class DiTBlock(nn.Module):
 
         x = x + gate_mlp * x_
         return x
+
+
+def modulate(x, shift, scale):
+    return x * (1 + scale) + shift
diff --git a/kimia_infer/models/tokenizer/whisper_Lv3/modeling_whisper.py b/kimia_infer/models/tokenizer/whisper_Lv3/modeling_whisper.py
index ed2b0f4..9e123af 100644
--- a/kimia_infer/models/tokenizer/whisper_Lv3/modeling_whisper.py
+++ b/kimia_infer/models/tokenizer/whisper_Lv3/modeling_whisper.py
@@ -27,21 +27,6 @@ import torch.utils.checkpoint
 from torch import nn
 from torch.nn import CrossEntropyLoss
 from torch.nn.utils.rnn import pad_sequence
-from flash_attn import (
-    flash_attn_qkvpacked_func,
-    flash_attn_func,
-    flash_attn_varlen_func,
-    flash_attn_varlen_qkvpacked_func,
-)
-
-try:
-    is_fa3_ready = True
-    from flash_attn_interface import (
-        flash_attn_varlen_func as fa3_varlen_func,
-        flash_attn_func as fa3_func,
-    )
-except:
-    is_fa3_ready = False
 
 from transformers.activations import ACT2FN
 from transformers.generation.logits_process import WhisperTimeStampLogitsProcessor
@@ -337,91 +322,44 @@ class WhisperAttention(nn.Module):
         hidden_states: [bsz, tgt_len, embed_dim]
         seq_lens: Optinal, [bsz]
         """
-        ### use flash_attn_func when seq_lens is None
-
-        query_states = self.q_proj(hidden_states)
+        # Standard attention implementation (replaced flash attention)
+        bsz, tgt_len, _ = hidden_states.size()
+        
+        # Get query, key, value projections
+        query_states = self.q_proj(hidden_states) * self.scaling
         key_states = self.k_proj(hidden_states)
         value_states = self.v_proj(hidden_states)
-        if cu_seqlens_q is None:
-            bsz, tgt_len, _ = hidden_states.size()
-            query_states = query_states.view(
-                bsz, tgt_len, self.num_heads, self.head_dim
-            )
-            key_states = key_states.view(bsz, tgt_len, self.num_heads, self.head_dim)
-            value_states = value_states.view(
-                bsz, tgt_len, self.num_heads, self.head_dim
-            )
-
-            if is_fa3_ready:
-                attn_output, attn_probs_ret = fa3_func(
-                    query_states,
-                    key_states,
-                    value_states,
-                    softmax_scale=self.scaling,
-                    causal=self.is_decoder,
-                )
-                if output_attentions:
-                    attn_probs = attn_probs_ret
-                else:
-                    attn_probs = None
-            else:
-                attn_output = flash_attn_func(
-                    query_states,
-                    key_states,
-                    value_states,
-                    dropout_p=self.dropout,
-                    softmax_scale=self.scaling,
-                    causal=self.is_decoder,
-                    return_attn_probs=output_attentions,
-                )
-                if output_attentions:
-                    attn_output, attn_probs = attn_output
-                else:
-                    attn_probs = None
-            attn_output = attn_output.view(bsz, tgt_len, self.embed_dim)
-        else:
-            query_states = query_states.view(-1, self.num_heads, self.head_dim)
-            key_states = key_states.view(-1, self.num_heads, self.head_dim)
-            value_states = value_states.view(-1, self.num_heads, self.head_dim)
-            if is_fa3_ready:
-                attn_output, attn_probs_ret = fa3_varlen_func(
-                    q=query_states,
-                    k=key_states,
-                    v=value_states,
-                    cu_seqlens_q=cu_seqlens_q,
-                    cu_seqlens_k=cu_seqlens_kv,
-                    max_seqlen_q=max_seqlen_q,
-                    max_seqlen_k=max_seqlen_kv,
-                    softmax_scale=self.scaling,
-                    causal=self.is_decoder,
-                )
-                if output_attentions:
-                    attn_probs = attn_probs_ret
-                else:
-                    attn_probs = None
-            else:
-                attn_output = flash_attn_varlen_func(
-                    q=query_states,
-                    k=key_states,
-                    v=value_states,
-                    cu_seqlens_q=cu_seqlens_q,
-                    cu_seqlens_k=cu_seqlens_kv,
-                    max_seqlen_q=max_seqlen_q,
-                    max_seqlen_k=max_seqlen_kv,
-                    dropout_p=self.dropout,
-                    softmax_scale=self.scaling,
-                    causal=self.is_decoder,
-                    return_attn_probs=output_attentions,
-                )
-                if output_attentions:
-                    attn_output, attn_probs = attn_output
-                else:
-                    attn_probs = None
-            attn_output = attn_output.view(-1, self.embed_dim)
-
-        ret = self.out_proj(attn_output)
-
-        return ret, attn_probs
+        
+        # Reshape for attention computation
+        query_states = query_states.view(bsz, tgt_len, self.num_heads, self.head_dim).transpose(1, 2)
+        key_states = key_states.view(bsz, tgt_len, self.num_heads, self.head_dim).transpose(1, 2)
+        value_states = value_states.view(bsz, tgt_len, self.num_heads, self.head_dim).transpose(1, 2)
+        
+        # Compute attention scores
+        attention_scores = torch.matmul(query_states, key_states.transpose(-1, -2))
+        
+        # Apply causal mask if needed
+        if self.is_decoder:
+            # Apply causal mask
+            causal_mask = torch.ones((tgt_len, tgt_len), device=hidden_states.device, dtype=torch.bool).triu(1)
+            causal_mask = causal_mask.unsqueeze(0).unsqueeze(0)  # [1, 1, tgt_len, tgt_len]
+            causal_mask = causal_mask.expand(bsz, self.num_heads, tgt_len, tgt_len)
+            mask_value = torch.finfo(attention_scores.dtype).min
+            mask_value = torch.tensor(mask_value, dtype=attention_scores.dtype)
+            attention_scores = attention_scores.masked_fill(causal_mask, mask_value)
+        
+        # Apply softmax and dropout
+        attention_probs = nn.functional.softmax(attention_scores, dim=-1)
+        attention_probs = nn.functional.dropout(attention_probs, p=self.dropout, training=self.training)
+        
+        # Apply attention to values
+        context = torch.matmul(attention_probs, value_states)
+        context = context.transpose(1, 2).contiguous().view(bsz, tgt_len, self.embed_dim)
+        
+        # Apply output projection
+        attn_output = self.out_proj(context)
+        
+        return attn_output, attention_probs if output_attentions else None
 
     # Copied from transformers.models.bart.modeling_bart.BartAttention.forward with BART->whisper
     def forward(
@@ -1279,8 +1217,8 @@ class WhisperDecoder(WhisperPreTrainedModel):
                 control over how to convert `input_ids` indices into associated vectors than the model's internal
                 embedding lookup matrix.
             output_attentions (`bool`, *optional*):
-                Whether or not to return the attentions tensors of all attention layers. See `attentions` under
-                returned tensors for more detail.
+                Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+                tensors for more detail.
             output_hidden_states (`bool`, *optional*):
                 Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors
                 for more detail.
diff --git a/requirements.txt b/requirements.txt
index 37c2a26..a998f3f 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -32,7 +32,7 @@ datasets
 torchdyn==1.0.6
 huggingface_hub
 loguru
-decord
+#decord
 blobfile
 timm
 sacrebleu==1.5.1

**Comment by yyaadet** [comment]
<img width="951" alt="Image" src="https://github.com/user-attachments/assets/68d2cae1-c77d-4338-9075-a0fc419cd488" />
bug 有点多，没法跑。

**Comment by MoyanZitto** [comment]
> <img alt="Image" width="951" src="https://private-user-images.githubusercontent.com/561840/437924176-68d2cae1-c77d-4338-9075-a0fc419cd488.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU3NTEzMTEsIm5iZiI6MTc0NTc1MTAxMSwicGF0aCI6Ii81NjE4NDAvNDM3OTI0MTc2LTY4ZDJjYWUxLWM3N2QtNDMzOC05MDc1LWEwZmM0MTljZDQ4OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyN1QxMDUwMTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iODYwMTVmZGZkZDA3ZjdkOWIzNTE0MjgwYWI2ZmIzNWRkZjBiZGY5ZDI4MjdiYWNiYTk5YjA5MzlkY2VlYjc4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.oe8kVB9fpBz9aEW6ePKCJBf71-k6MibJmqlo9Tc6UB0"> bug 有点多，没法跑。

@yyaadet sorry我们的问题，这里验证的不够充分，已经fix了
